### PR TITLE
FM2-395: Support standalone launch for Encounter context

### PR DESCRIPTION
**Pull request by theanandankit**
Issue Link: https://issues.openmrs.org/browse/FM2-395

### CODE REVIEW

**Comment by theanandankit** [comment]
@ibacher @VaishSiddharth PTAL.

File: omod/pom.xml [hunk 1]
```diff
@@ -85,6 +85,10 @@
 			<groupId>org.openmrs.module</groupId>
 			<artifactId>fhir2-omod</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>org.openmrs.module</groupId>
+			<artifactId>emrapi-api</artifactId>
```
Comment (ibacher) [inline omod/pom.xml:90]: This needs to be added to the required_modules in `config.xml`.

File: omod/src/main/java/org/openmrs/module/smartonfhir/page/controller/FindPatientPageController.java [hunk 2]
```diff
@@ -24,9 +24,10 @@
 public class FindPatientPageController {
 	
 	public void get(PageModel model, @RequestParam("app") AppDescriptor app, @RequestParam("token") String token,
-	        UiSessionContext sessionContext, UiUtils ui) throws UnsupportedEncodingException {
+	        @RequestParam("launchType") String launchType, UiSessionContext sessionContext, UiUtils ui)
+	        throws UnsupportedEncodingException {
 		model.addAttribute("afterSelectedUrl", app.getConfig().get("afterSelectedUrl").getTextValue() + "&token="
-		        + URLEncoder.encode(token, StandardCharsets.UTF_8.name()));
+		        + URLEncoder.encode(token, StandardCharsets.UTF_8.name()) + "&launchType=" + launchType);
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/smartonfhir/page/controller/FindPatientPageController.java]: Please URL encode the `launchType` as we do with the `token`

File: omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartVisitSelected.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+
+import org.keycloak.crypto.*;
+import org.keycloak.jose.jws.JWSBuilder;
+import org.keycloak.representations.JsonWebToken;
+import org.openmrs.module.smartonfhir.util.SmartSecretKeyHolder;
+
+public class SmartVisitSelected extends HttpServlet {
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartVisitSelected.java]: I'd really love it if we could collapse this functionality into a single servlet with some branching logic. Multiple servlets make the server lifecycle harder to manage (i.e., when this gets used in practice). E.g., replacing both this and SmartVisitSelected with:

```java
public class SmartLaunchOptionsSelected extends HttpServlet {
	
	public void doGet(HttpServletRequest req, HttpServletResponse res) throws IOException {
		String token = getParameter(req, "token");
		String patientId = getParameter(req, "patientId");
		String visitId = getParameter(req,"visitId");
		
		if (token == null || (patientId == null && visitId == null)) {
			res.sendError(HttpServletResponse.SC_BAD_REQUEST);
			return;
		}
		JsonWebToken tokenSentBack = new JsonWebToken();
		if (patientId != null) {
			tokenSentBack.setOtherClaims("patient", patientId);
		}
		
		if (visitId != null) {
			tokenSentBack.setOtherClaims("visit", visitId);
		}
		
		SecretKeySpec hmacSecretKeySpec = new SecretKeySpec(SmartSecretKeyHolder.getSecretKey(), JavaAlgorithm.HS256);
		KeyWrapper keyWrapper = new KeyWrapper();
		keyWrapper.setAlgorithm(Algorithm.HS256);
		keyWrapper.setSecretKey(hmacSecretKeySpec);
		SignatureSignerContext signer = new MacSignatureSignerContext(keyWrapper);
		
		String appToken = new JWSBuilder().jsonContent(tokenSentBack).sign(signer);
		String encodedToken = URLEncoder.encode(appToken, StandardCharsets.UTF_8.name());
		
		String decodedUrl = URLDecoder.decode(token, StandardCharsets.UTF_8.name());
		res.sendRedirect(decodedUrl.replace("{APP_TOKEN}", encodedToken));
	}
	
	private String getParameter(HttpServletRequest request, String parameter) {
		String result = request.getParameter(parameter);
		if (result == null || result.isEmpty()) {
			return null;
		}
		
		return  result;
	}
}
```

File: omod/src/main/webapp/pages/findVisit.gsp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                                                                 bFilter: true,
+                                                                 bJQueryUI: true,
+                                                                 bLengthChange: false,
+                                                                 iDisplayLength: 10,
+                                                                 sPaginationType: '\"full_numbers\"',
+                                                                 bSort: false,
+                                                                 sDom: '\'ft<\"fg-toolbar ui-toolbar ui-corner-bl ui-corner-br ui-helper-clearfix datatables-info-and-pg \"ip>\''
+                                                         ]
+]) }
+<% } %>
```
Comment (ibacher) [inline omod/src/main/webapp/pages/findVisit.gsp]:
```suggestion
<% } %>

```

File: omod/src/main/resources/apps/findVisit_app.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  {
+    "id": "smartonfhir.search.visit",
+    "description": "List of all visits for a particular patient",
+    "order": 2,
+    "config": {
+      "afterSelectedUrl": "ms/smartVisitSelected?visitId={{visitId}}&patientId={{patient.uuid}}&token={{token}}",
+      "heading": "Select Visit"
+    }
+  }
+]
```
Comment (ibacher) [inline omod/src/main/resources/apps/findVisit_app.json]:
```suggestion
]

```

**Review by ibacher — COMMENTED** [review]
@theanandankit Thanks for this. A few small things to tweak and it should be fine.

File: omod/pom.xml (hunk 1 above)
Comment (theanandankit) [reply omod/pom.xml:90]: I will add this.

File: omod/src/main/java/org/openmrs/module/smartonfhir/page/controller/FindPatientPageController.java (hunk 2 above)
Comment (theanandankit) [reply omod/src/main/java/org/openmrs/module/smartonfhir/page/controller/FindPatientPageController.java]: Thanks, I will encode this too.

File: omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartVisitSelected.java (hunk 3 above)
Comment (theanandankit) [reply omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartVisitSelected.java]: Yes, you are right we can merge this servlet with the already existing SmartPatientSelected servlet. I will do this too.

**Comment by theanandankit** [comment]
@ibacher I have merged the new servlet into an already existing servlet and renamed it to `SmartLaunchOptionServlet`. Also made all required changes. PTAL

**Comment by theanandankit** [comment]
@ibacher PTAL.

**Review by ibacher — APPROVED** [review]

**Pull request merged**